specs/language/basic.tex

```diff
@@ -1,5 +1,113 @@
 \Ch{Basic Concepts}{Basic}
 
+\Sec{Types}{Basic.types}
+
+\p The \textit{object representation} of an object of type \texttt{T} is the
+sequence of \textit{N} bytes taken up by the object of type \texttt{T}, where
+\textit{N} equals \texttt{sizeof(T)}\footnote{\texttt{sizeof(T)} returns the
+size of the object as-if it's stored in device memory, and determining the size
+if it's stored in another memory space is not possible.}. The \textit{object
+representation} of an object may be different based on the \textit{memory space}
+it is stored in (\ref{Intro.Memory.Spaces}).
+
+\p The \textit{value representation} of an object is the set of bits that hold
+the value of type \texttt{T}. Bits in the object representation that are not
+part of the value representation are \textit{padding bits}.
+
+\p An \textit{object type} is a type that is not a function type, not a
+reference type, and not a void type.
+
+\p A \textit{class type} is a data type declared with either the \texttt{class}
+or \texttt{struct} keywords (\ref{Classes}). A class type \texttt{T} may be
+declared as incomplete at one point in a translation unit via a \textit{forward
+declaration}, and complete later with a full definition. The type \texttt{T} is
+the same type throughout the translation unit.
+
+\p There are special implementation-defined types such as \textit{handle types},
+which fall into a category of \textit{standard intangible types}. Intangible
+types are types that have no defined object representation or value
+representation, as such the size is unknown at compile time.
+% Note: The above definition is likely incomplete, and it is unclear if minimum
+% precision types should be intangible.
+
+\p A class type \texttt{T} is an \textit{intangible class type} if it contains
+an base classes or members of intangible class type, standard intangible type,
+or arrays of such types. Standard intangible types and intangible class types
+are collectively called \textit{intangible types}(\ref{Intangible}).
+
+\p An object type is an \textit{incomplete type} if the compiler lacks
+sufficient information to determine the size of an object of type \texttt{T},
+and it is not an intangible type. It is a \textit{complete type} if the compiler
+has sufficient information to determine the size of an object of type
+\texttt{T}, or if the type is known to be an intangible type. An object may not
+be defined to have an \textit{incomplete} type.
+
+\p Arithmetic types (\ref{Basic.types.arithmetic}), enumeration types, and
+\textit{cv-qualified} versions of these types are collectively called
+\textit{scalar types}.
+
+\p Vectors of scalar types declared with the built-in \texttt{vector<T,N>}
```

Comment:
Presumably somewhere we have something like `use float4 = vector<float, 4>`?  Wondering if it'd be helpful to mention these names that most users of HLSL are likely more familiar with here?

Reply:
Yea, we have the typedefs in clang [here](https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/hlsl/hlsl_basic_types.h). We use `typedef` rather than `using` because `using` is an HLSL 2021 feature, so that makes the header compatible to older versions.

The C and C++ specs defines all the typedefs later in the document, which had been my plan. Basically anything that can be implemented in a header goes way late.

Reply:
I've filed an issue to track adding this documentation:
https://github.com/microsoft/hlsl-specs/issues/203

Reply:
I'm going to mark this as resolved for this thread. The additional documentation will be handled in a separate PR.